crypto/s2n_hash.c

```diff
@@ -39,28 +45,65 @@ bool s2n_hash_evp_fully_supported()
     return s2n_use_evp_impl() && !s2n_use_custom_md5_sha1();
 }
 
-const EVP_MD *s2n_hash_alg_to_evp_md(s2n_hash_algorithm alg)
+S2N_RESULT s2n_hash_algorithms_init()
 {
-    switch (alg) {
-        case S2N_HASH_MD5:
-            return EVP_md5();
-        case S2N_HASH_SHA1:
-            return EVP_sha1();
-        case S2N_HASH_SHA224:
-            return EVP_sha224();
-        case S2N_HASH_SHA256:
-            return EVP_sha256();
-        case S2N_HASH_SHA384:
-            return EVP_sha384();
-        case S2N_HASH_SHA512:
-            return EVP_sha512();
-#if defined(S2N_LIBCRYPTO_SUPPORTS_EVP_MD5_SHA1_HASH)
-        case S2N_HASH_MD5_SHA1:
-            return EVP_md5_sha1();
+#if S2N_LIBCRYPTO_SUPPORTS_PROVIDERS
```

Comment:
Doesn't this need to check for defined?
```suggestion
#ifdef S2N_LIBCRYPTO_SUPPORTS_PROVIDERS
```

Though `s2n_hash_test` should be testing MD5/SHA1, which should be failing for FIPS without this working, right?

Reply:
Honestly, either way works. Our feature tests set S2N_LIBCRYPTO_SUPPORTS_PROVIDERS=1 when providers are available, and don't set S2N_LIBCRYPTO_SUPPORTS_PROVIDERS at all when they're not. The only difference would be that this would fail if something set S2N_LIBCRYPTO_SUPPORTS_PROVIDERS=0, which I lean towards being correct behavior.

Reply:
https://godbolt.org/z/6KEEj9Edn

Reply:
Ahh interesting ok, yeah that makes sense to check if it's true then. Do you know where we set it equal to 1? I was looking here and thought it was just defined generically:
https://github.com/aws/s2n-tls/blob/4c5e35afc47f9218f0f75d1d4902a70974c9b33d/CMakeLists.txt#L312

Reply:
Oh I guess adding a definition just defines it set to 1?

Reply:
We poked around. It looks like the "1" default for "-D" isn't part of the C standard, but has been the de-factor standard basically since C compilers existed. It's definitely true for gcc and clang.

But as a follow-up, since this isn't our only #if instead of #ifdef, I'm going to change the build to set an explicit default.